ecs-agent/acs/session/payload_responder.go

```diff
@@ -0,0 +1,162 @@
+		ackFunc func(*ecsacs.AckRequest, []*ecsacs.IAMRoleCredentialsAckRequest)) error
+}
+
+// payloadResponder implements the wsclient.RequestResponder interface for responding
+// to ecsacs.PayloadMessage messages sent by ACS.
+type payloadResponder struct {
+	taskMessageHandler TaskMessageHandler
+	respond            wsclient.RespondFunc
+}
+
+// NewPayloadResponder returns an instance of the payloadResponder struct.
+func NewPayloadResponder(taskMessageHandler TaskMessageHandler,
+	responseSender wsclient.RespondFunc) wsclient.RequestResponder {
+	r := &payloadResponder{
+		taskMessageHandler: taskMessageHandler,
+	}
+	r.respond = ResponseToACSSender(r.Name(), responseSender)
+	return r
+}
+
+func (*payloadResponder) Name() string { return "payload responder" }
+
+func (r *payloadResponder) HandlerFunc() wsclient.RequestHandler {
+	return r.handlePayloadMessage
+}
+
+func (r *payloadResponder) handlePayloadMessage(message *ecsacs.PayloadMessage) {
+	logger.Debug(fmt.Sprintf("Handling %s", PayloadMessageName))
+
+	// Validate fields in the message.
+	if err := validatePayloadMessage(message); err != nil {
+		logger.Error(fmt.Sprintf("Error validating %s received from ECS", PayloadMessageName), logger.Fields{
+			field.Error: err,
+		})
+		return
+	}
+
+	// Handle payload message.
+	err := r.taskMessageHandler.ProcessTaskMessage(message, r.ackFunc)
+	if err != nil {
+		logger.Critical(fmt.Sprintf("Unable to handle %s", PayloadMessageName), logger.Fields{
+			field.MessageID: aws.StringValue(message.MessageId),
+			field.Error:     err,
+		})
+	}
+}
+
+// ackFunc sends ACKs of the payload message and of the credentials associated with the tasks contained in the payload
+// message.
+func (r *payloadResponder) ackFunc(payloadAck *ecsacs.AckRequest, credsAcks []*ecsacs.IAMRoleCredentialsAckRequest) {
+	go r.sendAck(payloadAck)
+
+	for _, credsAck := range credsAcks {
+		go r.sendAck(credsAck)
+	}
+}
+
+// sendAck handles the sending of an individual specific ACK, assuming it is of type
+// ecsacs.IAMRoleCredentialsAckRequest or ecsacs.AckRequest.
+func (r *payloadResponder) sendAck(ackRequest interface{}) {
+	var credentialsAck *ecsacs.IAMRoleCredentialsAckRequest
+	var payloadMessageAck *ecsacs.AckRequest
+	credentialsAck, ok := ackRequest.(*ecsacs.IAMRoleCredentialsAckRequest)
+	if ok {
+		logger.Debug(fmt.Sprintf("ACKing credentials associated with %s", PayloadMessageName), logger.Fields{
+			"credentialsID": aws.StringValue(credentialsAck.CredentialsId),
+			field.MessageID: aws.StringValue(credentialsAck.MessageId),
+		})
+	} else {
+		payloadMessageAck, ok = ackRequest.(*ecsacs.AckRequest)
+		if ok {
+			logger.Debug(fmt.Sprintf("ACKing %s", PayloadMessageName), logger.Fields{
+				field.MessageID: aws.StringValue(payloadMessageAck.MessageId),
+			})
+		} else {
+			logger.Error(fmt.Sprintf("Error sending acknowledgement: %s",
+				"ackRequest does not hold type ecsacs.IAMRoleCredentialsAckRequest or ecsacs.AckRequest"))
+			return
+		}
+	}
```

Comment:
Not a big fan of if-else's. Is it possible to make this a switch statement?

Reply:
I agree that we should use switch over if-else if there is no other advantage to using one over the other.

In this case, my thinking is that if we use a switch statement, I believe we'll be required to potentially do an unnecessary type assertion upfront (i.e., we would need to execute both 
`credentialsAck, ok := ackRequest.(*ecsacs.IAMRoleCredentialsAckRequest)` and `payloadMessageAck, ok = ackRequest.(*ecsacs.AckRequest)`
before calling the switch statement. Whereas with this if else, `payloadMessageAck, ok = ackRequest.(*ecsacs.AckRequest)` is only executed if necessary.

Perhaps I may be missing some other way of approaching this using a switch statement that doesn't require an unnecessary type assertion. Is there any that comes to mind for you?

Reply:
Hmm thats right. I am not able to come up with another way to handle it. We can leave it as it is then. Thanks.